### PR TITLE
feat(string-formatter): add string formatter

### LINKED ISSUE #24: `StringFormatter`
### Context

Absorb `StringFormatter` pattern.

### PR DESCRIPTION
**Pull request by SimonFinney**
## Affected issues

- Resolves #24 

## Proposed changes

- Add `StringFormatter` component

## Testing instructions

### CODE REVIEW

File: src/components/StringFormatter/StringFormatter.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  /** @type {string} Optional class name. */
+  className: string,
+
+  /** @type {boolean} Whether or not the value should be truncated. */
+  truncate: bool,
+
+  /** @type {string} Value to format. */
+  value: string.isRequired,
+
+  /** @type {string} width to max the string at. */
```
Comment (jendowns) [inline src/components/StringFormatter/StringFormatter.js]: Small thing, but can we reword this to: "Max width of string."

File: src/components/StringFormatter/StringFormatter.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  width: string,
+};
+
+StringFormatter.defaultProps = {
+  className: null,
```
Comment (jendowns) [inline src/components/StringFormatter/StringFormatter.js]: I think for almost all other component we use an empty string in this case. 🤔 Could this be `className: ''` for consistency?

File: src/components/StringFormatter/__tests__/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  beforeEach(() => {
+    stringFormatter = shallow(<StringFormatter {...props} />);
+  });
+
+  it('renders', () => {
+    expect(stringFormatter).toMatchSnapshot();
+  });
+
+  it('renders the truncation variation', () => {
```
Comment (jendowns) [inline src/components/StringFormatter/__tests__/index.js]: "renders the **truncated** variation" ?

File: src/components/StringFormatter/_index.scss [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+/// @copyright IBM Security 2019
+////
+
+@import '../Component/mixins';
+
+@include component($name: string-formatter) {
+  display: inline-block;
+
+  &--truncate {
+    max-width: 80ch;
```
Comment (jendowns) [inline src/components/StringFormatter/_index.scss]: I realize we are dealing with a character width here, but that `80` is still divisible by the mini unit 🤔 Could you use mini-units here for the value?

File: src/components/StringFormatter/_mocks_/index.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/**
```
Comment (jendowns) [inline src/components/StringFormatter/_mocks_/index.js]: I'm wondering if it is worth having this file, if it's just `value: 'Value'` being saved? Could we simplify and remove the mock file?

File: src/components/StringFormatter/StringFormatter.js (hunk 2 above)
Comment (SimonFinney) [reply src/components/StringFormatter/StringFormatter.js]: Good catch!

File: src/components/StringFormatter/__tests__/index.js (hunk 3 above)
Comment (SimonFinney) [reply src/components/StringFormatter/__tests__/index.js]: Thanks!

File: src/components/StringFormatter/_index.scss [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+@import '@carbon/layout/scss/mini-unit';
+
+@import '../Component/mixins';
+
+@include component($name: string-formatter) {
+  display: inline-block;
+
+  &--truncate {
+    max-width: carbon--mini-units($count: 84);
```
Comment (jendowns) [inline src/components/StringFormatter/_index.scss]: I know I suggested changing this but now I'm questioning it. 😅 Since it comes out as `42rem` which is a bit different from what the original intent was, using the `ch` unit.
